k8s/src/main/scala/io/buoyant/k8s/resources.scala

```diff
@@ -0,0 +1,308 @@
+
+  protected def infiniteRetryFilter = new RetryFilter[http.Request, http.Response](
+    RetryPolicy.backoff(backoffs) {
+      // We will assume 5xx are retryable, everything else is not for now
+      case (_, Return(rep)) => rep.status.code >= 500 && rep.status.code < 600
+      case (_, Throw(NonFatal(ex))) =>
+        log.error(s"retrying k8s endpoints request on error $ex")
+        true
+    },
+    HighResTimer.Default,
+    stats,
+    RetryBudget.Infinite
+  )
+
+  /**
+   * Watch this resource for changes, using a chunked HTTP request.
+   *
+   * TODO: investigate k8s websockets support as an alternative HTTP approach
+   */
+  def watch(
+    labelSelector: Option[String] = None,
+    fieldSelector: Option[String] = None,
+    resourceVersion: Option[String] = None
+  ): (AsyncStream[W], Closable) = {
+    val close = new AtomicReference[Closable](Closable.nop)
+
+    // Internal method used to recursively retry watches as needed on failures.
+    def _watch(resourceVersion: Option[String] = None): AsyncStream[W] = {
+      val req = Api.mkreq(http.Method.Get, path, None,
+        "watch" -> Some("true"),
+        "labelSelector" -> labelSelector,
+        "fieldSelector" -> fieldSelector,
+        "resourceVersion" -> resourceVersion)
+
+      def currentVersion(initial: Option[String], watch: AsyncStream[W]): Future[Option[String]] = {
+        watch.foldLeft(initial) {
+          (version: Option[String], watch: W) => watch.resourceVersion
+        }
+      }
+
+      def watchFromResponse(rsp: Response): AsyncStream[W] = rsp.status match {
+        // NOTE: 5xx-class statuses will be retried by the infiniteRetryFilter above.
+        case http.Status.Ok =>
+          close.set(Closable.make { _ =>
+            log.debug("k8s watch closed")
+            Future {
+              rsp.reader.discard()
+            } handle {
+              case _: Reader.ReaderDiscarded =>
+            }
+          })
+          val theStream = Json.readStream[W](rsp.reader, Api.BufSize)
+          theStream ++
+            AsyncStream.fromFuture(
+              currentVersion(resourceVersion, theStream) flatMap {
+                case Some(version) =>
+                  Future.value(_watch(Some(version)))
+                case None =>
+                  // In this case, we want to try loading the initial information instead before watching again.
+                  restartWatches(labelSelector, fieldSelector).map {
+                    case (ws, ver) =>
+                      AsyncStream.fromSeq(ws) ++
+                        _watch(ver)
+                  }
+              }
+            ).flatten
+
+        case status =>
+          close.set(Closable.nop)
+          log.debug("k8s failed to watch endpoints: %d %s", status.code, status.reason)
+          val f = Future.exception(Api.UnexpectedResponse(rsp))
+          AsyncStream.fromFuture(f)
+      }
+
+      val retryingClient = infiniteRetryFilter andThen client
```

Comment:
looks like we're about half/half for infix vs. dots on `andThen`, and finagle uses infix most of the time.